File: hackerrank/problem solving/algorithms/2 - implementation/53 - flatland space stations/with morfo library/flatland_space_stations.cs
// https://www.hackerrank.com/challenges/flatland-space-stations/problem?isFullScreen=true

using Morfo.IO;

int nCities = Reader.ReadLn<int, int[]>()[0];
int lastCity = nCities - 1;
int[] stations = Reader.ReadLn<int, int[]>();
Array.Sort(stations);
Console.WriteLine(MaxDistanceFromSpaceStation(lastCity, stations));

// n: length of stations
// T: O(n)
// S: O(1) extra space
int MaxDistanceFromSpaceStation(int lastCity, int[] stations)
{
    int maxDist = stations[0];

    foreach (var (prev, curr) in stations.Zip(stations.Skip(1)))
    {
        int gap = (curr - prev) / 2;
        maxDist = Math.Max(maxDist, gap);
    }

    return Math.Max(maxDist, lastCity - stations[^1]);
}

File: hackerrank/problem solving/algorithms/2 - implementation/53 - flatland space stations/with morfo library/flatland_space_stations.js
// https://www.hackerrank.com/challenges/flatland-space-stations/problem?isFullScreen=true

import { idiv, readLn, sort } from 'morfo';

const [nCities, _] = readLn(Number);
const lastCity = nCities - 1;
const stations = sort(readLn(Number));
console.log(maxDistanceFromSpaceStation(lastCity, stations));

// n: length of stations
// T: O(n)
// S: O(1) extra space
function maxDistanceFromSpaceStation(lastCity, stations) {
  let maxDist = stations[0];

  for (let i = 1; i < stations.length; i++) {
    const gap = idiv(stations[i] - stations[i - 1], 2);
    maxDist = Math.max(maxDist, gap);
  }

  return Math.max(maxDist, lastCity - stations.at(-1));
}
